kanidmd/src/lib/tracing_tree/subscriber.rs

```diff
@@ -0,0 +1,454 @@
+    pub name: &'static str,
+    pub processed_buf: Vec<TreeProcessed<E>>,
+    pub uuid: Option<String>,
+    pub out: TreeIo,
+    pub nested_duration: u64,
+    pub total_duration: u64,
+}
+
+pub(crate) enum TreeProcessed<E> {
+    Event(TreeEvent<E>),
+    Span(TreeSpanProcessed<E>),
+}
+
+impl<E: EventTagSet> TreeSubscriber<E> {
+    // Only reason this is public is so we can configure at runtime.
+    pub fn new(fmt: LogFmt, log_tx: UnboundedSender<TreeProcessor<E>>) -> Self {
+        TreeSubscriber {
+            inner: Registry::default().with(TreeLayer { fmt, log_tx }),
+        }
+    }
+
+    // These are the preferred constructors.
+
+    #[allow(dead_code)]
+    pub fn json(log_tx: UnboundedSender<TreeProcessor<E>>) -> Self {
+        TreeSubscriber::new(LogFmt::Json, log_tx)
+    }
+
+    pub fn pretty(log_tx: UnboundedSender<TreeProcessor<E>>) -> Self {
+        TreeSubscriber::new(LogFmt::Pretty, log_tx)
+    }
+}
+
+impl<E: EventTagSet> Subscriber for TreeSubscriber<E> {
+    fn enabled(&self, metadata: &Metadata) -> bool {
+        self.inner.enabled(metadata)
+    }
+
+    fn max_level_hint(&self) -> Option<tracing::metadata::LevelFilter> {
+        self.inner.max_level_hint()
+    }
+
+    fn new_span(&self, span: &Attributes) -> Id {
+        self.inner.new_span(span)
+    }
+
+    fn record(&self, span: &Id, values: &Record) {
+        self.inner.record(span, values)
+    }
+
+    fn record_follows_from(&self, span: &Id, follows: &Id) {
+        self.inner.record_follows_from(span, follows)
+    }
+
+    fn event(&self, event: &Event) {
+        self.inner.event(event)
+    }
+
+    fn enter(&self, span: &Id) {
+        self.inner.enter(span)
+    }
+
+    fn exit(&self, span: &Id) {
+        self.inner.exit(span)
+    }
+
+    fn clone_span(&self, id: &Id) -> Id {
+        self.inner.clone_span(id)
+    }
+
+    fn try_close(&self, id: Id) -> bool {
+        self.inner.try_close(id)
+    }
+}
+
+impl<E: EventTagSet> TreeLayer<E> {
+    fn log_to_parent(&self, logs: Tree<E>, parent: Option<SpanRef<Registry>>) {
+        match parent {
+            // The parent exists- write to them
+            Some(span) => span
+                .extensions_mut()
+                .get_mut::<TreeSpan<E>>()
+                .expect("Log buffer not found, this is a bug")
+                .log(logs),
+            // The parent doesn't exist- send to formatter
+            None => self
+                .log_tx
+                .send(TreeProcessor {
+                    fmt: self.fmt,
+                    logs,
+                })
+                .expect("Processing channel has been closed, cannot log events."),
+        }
+    }
+
+    fn alarm(event: &TreeEvent<E>, maybe_scope: Option<ScopeFromRoot<Registry>>) -> fmt::Result {
+        // This is an emergency and should be sent to the admin immediately
+        // Hence why we are formatting in the working thread
+        let mut writer = event.timestamp.to_rfc3339();
```

Comment:
What type is writer here? I'm curious how this works :)

Reply:
It's just a `String`, I did it like that since `to_rfc3339` returns a `String` and I would've appended it to an empty `String` anyways. I'm going to make this clearer in the next commit, though.